packages/@aws-cdk/aws-ec2/lib/cfn-init.ts

```diff
@@ -0,0 +1,387 @@
+   *
+   * @default '000644'
+   */
+  readonly mode?: string;
+}
+
+/**
+ * Create files on the EC2 instance.
+ */
+export class InitFile extends InitElement {
+  /**
+   * Use a literal string as the file content
+   */
+  public static fromString(fileName: string, s: string, options: InitFileOptions = {}): InitFile {
+  }
+
+  /**
+   * Write a symlink with the given symlink target
+   */
+  public static symlink(fileName: string, target: string, options: InitFileOptions = {}): InitFile {
+  }
+
+  /**
+   * Use a base64 string as the file content
+   */
+  public static fromBase64String(fileName: string, base64: string, options: InitFileOptions = {}): InitFile {
+  }
+
+  /**
+   * Use a JSON object as the file content, write it to a JSON file.
+   *
+   * May contain tokens.
+   */
+  public static fromJsonObject(fileName: string, obj: Record<string, any>, options: InitFileOptions = {}): InitFile {
+  }
+
+  /**
+   * Read a file from disk and use its contents
+   *
+   * The file will be embedded in the template, so care should be taken to not
+   * exceed the template size.
+   *
+   * The file will be base64-encoded if it contains nonprintable characters.
+   */
+  public static fromFileInline(targetFileName: string, sourceFileName: string, options: InitFileOptions = {}): InitFile {
+  }
+
+  /**
+   * Create an asset from the given file and use that
+   *
+   * This is appropriate for files that are too large to embed into the template.
+   */
+  public static fromFileAsset(targetFileName: string, sourceFileName: string, options: InitFileOptions = {}): InitFile {
+  }
+
+  /**
+   * Download from a URL at instance startup time
+   */
+  public static fromUrl(targetFileName: string, url: string, options: InitFileOptions = {}): InitFile {
+  }
+
+  /**
+   * Use a file from an asset at instance startup time
+   */
+  public static fromAsset(targetFileName: string, asset: s3_assets.Asset, options: InitFileOptions = {}): InitFile {
+  }
+
+  protected constructor() {
+    super();
+  }
+}
+
+/**
+ * Create Linux/UNIX groups and to assign group IDs.
+ *
+ * Not supported for Windows systems.
+ */
+export class InitGroup extends InitElement {
+  /**
+   * Map a group name to a group id
+   */
+  public static fromName(groupName: string, gid: number): InitGroup {
+  }
+
+  protected constructor() {
+    super();
+  }
+}
+
+/**
+ * Create Linux/UNIX users and to assign user IDs.
+ *
+ * Users are created as non-interactive system users with a shell of
+ * /sbin/nologin. This is by design and cannot be modified.
+ *
+ * Not supported for Windows systems.
+ */
+export class InitUser extends InitElement {
+  /**
+   * Map a user name to a user id
+   */
+  public static fromName(userName: string, uid: number, groups?: string[], homeDir?: string): InitUser {
+  }
+
+  protected constructor() {
+    super();
+  }
+}
+
+/**
+ * A package to be installed during cfn-init time
+ */
+export class InitPackage extends InitElement {
```

Comment:
Is this one of those examples where we'd want to provide an extension point for when a new repo type is added? Something like:

```ts
public static custom(repositoryName: string, packageName: string, ...versions: string[]): InitPackage
```

I'm thinking of use cases like https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-cognito/lib/user-pool-client.ts#L96-L133; for example, if CloudFormation starts supporting npm.

Reply:
Not opposed to it.